File: smart-farm-server/src/models/Device.js
const mongoose = require('mongoose');

const DeviceSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    location: { type: String, default: '', trim: true },
    status: { type: String, enum: ['online', 'offline'], default: 'offline', index: true },
    ownerId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    firmwareVersion: { type: String, default: '' },
    // External identifier reported by device over MQTT (e.g., esp32-7C87CE30CCCC)
    externalId: { type: String, unique: true, sparse: true, index: true },
    // Simple automation thresholds
    autoPumpEnabled: { type: Boolean, default: false },
    autoPumpSoilBelow: { type: Number }, // percentage
    autoFanEnabled: { type: Boolean, default: false },
    autoFanTempAbove: { type: Number }, // Celsius
  autoLightEnabled: { type: Boolean, default: false },
  autoLightLuxBelow: { type: Number }, // lux threshold
  // Hysteresis values
  autoFanHysteresis: { type: Number }, // °C
  autoPumpHysteresis: { type: Number }, // %
  autoLightHysteresis: { type: Number }, // lux
  // Minimum interval between toggles to avoid flapping (seconds)
  minToggleIntervalSec: { type: Number, default: 0 },
  // Last toggle timestamps
  lastFanToggleAt: { type: Date },
  lastPumpToggleAt: { type: Date },
  lastLightToggleAt: { type: Date },
  lastFanState: { type: String, enum: ['ON','OFF'] },
  lastPumpState: { type: String, enum: ['ON','OFF'] },
  lastLightState: { type: String, enum: ['ON','OFF'] },
  // Presence tracking
  lastSeenAt: { type: Date },
  // Simple daily schedule (single slot per target)
  schedFanOn: { type: String }, // HH:MM
  schedFanOff: { type: String },
  schedFanDays: { type: String }, // comma separated 0-6
  schedLightOn: { type: String },
  schedLightOff: { type: String },
  schedLightDays: { type: String },
  schedPumpOn: { type: String },
  schedPumpOff: { type: String },
  schedPumpDays: { type: String },
    createdAt: { type: Date, default: Date.now },
  },
  { minimize: true }
);

module.exports = mongoose.model('Device', DeviceSchema);

File: code_frontend/lib/models/device.dart
class Device {
  final String id;
  final String name;
  final String? location;
  final String? externalId;
  final String? status;
  final String? opMode;
  final bool autoFanEnabled;
  final bool autoPumpEnabled;
  final bool autoLightEnabled;
  final String? lastFanState;
  final String? lastLightState;
  final String? lastPumpState;
  final String? pairedSensorId;

  Device({
    required this.id,
    required this.name,
    this.location,
    this.externalId,
    this.status,
    this.opMode,
    this.autoFanEnabled = false,
    this.autoPumpEnabled = false,
    this.autoLightEnabled = false,
    this.lastFanState,
    this.lastLightState,
    this.lastPumpState,
    this.pairedSensorId,
  });

  factory Device.fromJson(Map<String, dynamic> j) => Device(
    id: j['_id'] ?? j['id'] ?? '',
    name: j['name'] ?? '',
    location: j['location'],
    externalId: j['externalId'],
    status: j['status'],
    opMode: j['opMode'],
    autoFanEnabled: j['autoFanEnabled'] ?? false,
    autoPumpEnabled: j['autoPumpEnabled'] ?? false,
    autoLightEnabled: j['autoLightEnabled'] ?? false,
    lastFanState: j['lastFanState']?.toString(),
    lastLightState: j['lastLightState']?.toString(),
    lastPumpState: j['lastPumpState']?.toString(),
    pairedSensorId: j['pairedSensorId']?.toString(),
  );

  Device copyWith({
    String? id,
    String? name,
    String? location,
    String? externalId,
    String? status,
    String? opMode,
    bool? autoFanEnabled,
    bool? autoPumpEnabled,
    bool? autoLightEnabled,
    String? lastFanState,
    String? lastLightState,
    String? lastPumpState,
    String? pairedSensorId,
  }) => Device(
    id: id ?? this.id,
    name: name ?? this.name,
    location: location ?? this.location,
    externalId: externalId ?? this.externalId,
    status: status ?? this.status,
    opMode: opMode ?? this.opMode,
    autoFanEnabled: autoFanEnabled ?? this.autoFanEnabled,
    autoPumpEnabled: autoPumpEnabled ?? this.autoPumpEnabled,
    autoLightEnabled: autoLightEnabled ?? this.autoLightEnabled,
    lastFanState: lastFanState ?? this.lastFanState,
    lastLightState: lastLightState ?? this.lastLightState,
    lastPumpState: lastPumpState ?? this.lastPumpState,
    pairedSensorId: pairedSensorId ?? this.pairedSensorId,
  );
}
